### DISCUSSION TITLE
Express compatibility layer

### DISCUSSION THREAD

**Discussion by endel** [uNetworking/uWebSockets.js#565] [Show and tell]
Hello everyone 👋 

I've dedicated this week fully to develop an Express compatibility layer on top of uWebSockets.js, and made it available here: https://github.com/endel/uWebSockets-express

With this compatibility layer, you can use the familiar Express API's while leveraging uWebSockets.js under the hood. 

**Example:**

```typescript
import uWS from "uWebSockets.js"
import expressify from "uwebsockets-express"
import serveIndex from "serve-index";

const uwsApp = uWS.App();
const app = expressify(uwsApp);

// use existing middleware implementations!
app.use('/', serveIndex(path.join(__dirname, ".."), { icons: true, hidden: true }))
app.use('/', express.static(path.join(__dirname, "..")));

app.get("/hello", (req, res) => {
  res.json({ hello: "world!" });
});
```

I've made some initial tests and some existing express middlewares are already working with this compatibility layer, such as [express-session](https://github.com/expressjs/session), [serve-index](https://github.com/expressjs/serve-index), and [serve-static](https://github.com/expressjs/serve-static). 

If you can try it out, provide feedback, and/or report bugs regarding the usage on well-known middlewares it would be really helpful to improve and continue the development of this compatibility layer!

(This compatibility layer is going to be available as part of the upcoming uWebSockets.js support on [Colyseus](https://github.com/colyseus/colyseus) 🎉 )

Cheers, hope you enjoy!

**Comment by e3dio** [comment]
Lots of people try to make Express layer for this library, fine to try but my opinion is no point adding large complex layer that is not necessary and gives false impression that it is compatible with Express when is not, best to just do standard practice ways of doing simple things and use 50x less code that is readable with zero bugs etc, Your source is 440 lines of code which all need to be read through and overcomplicate, not to mention the performance hit whatever is happening every request, and you are doing strange things like what is this https://github.com/endel/uWebSockets-express/blob/master/src/Request.ts#L73

**Reply by joshxyzhimself** [reply]
Yep there's also nanoexpress

https://github.com/nanoexpress/nanoexpress

I also wrote a (lame unreleased) similar abstraction for personal use to have ease with regards to:

- async functions
- parsing of json requests
- content-type assignment (text, html, json, octet, or whatever for file)
- static files
- compression (gzip, brotli)
- cache-control
- redirects
- error handling

e3dio, performance wise you're hella correct but it's a nice tradeoff in terms of readability and not repeating code.

**Reply by endel** [reply]
Hi @joshxyzhimself, would you mind sharing your implementation so we could combine the solutions somehow? I'm curious about your approach to async functions, cache-control, and error handling!

I did see the nanoexpress project - but they do a lot more than I need. I only need a thin and simple layer that mimics express while actually using uWebSockets.

**Reply by joshxyzhimself** [reply]
Hi, yes it's now here: https://github.com/joshxyzhimself/uwu

I haven't implemented middlewares, but simple usage of it looks like this:

```js
// serve_handler here returns an internal handler function
app.post('/some-path', uwu.serve_handler(async (response, request) => {

  // application/json sent by client requests are accessible here
  console.log(request.json);

  // this lets me send json responses easily, internally stringifies it then sets content-type
  response.json = { foo: 'bar' };

  // if i want "no-store, max-age=0", say for sensitive data
  response.headers['Cache-Control'] = uwu.cache_control_types.no_store;

  // if i want any compressions for gzip and brotli
  response.compress = true;
});
```

On error handling we use and abuse `assert` a lot to catch avoidable uws dumps (e.g. when someone sets response.buffer to null, internal call of res.end(null) causes dump); use of typescript's `// @ts-check` also helps a lot in that.

On request-level errors we emit to an event emitter so we can pipe them somewhere (e.g. to console, file, or any log aggregator depending on environment) and get to review them when needed.

Keep in mind it could be clunky and buggy lol.
